File: windows/src/tokenizer.js
// Port of VoxCore/Parsing/Tokenizer.swift. Keep behaviour identical: shared/grammar-cases.json checks both.

const WORD_CHAR = /[\p{L}\p{N}'’]/u;

/** Splits on anything that is not a letter, digit or apostrophe; lowercases; drops apostrophes. */
export function tokenize(text) {
  const tokens = [];
  let start = -1;
  const chars = Array.from(text); // code points, like Swift Characters for our purposes
  let offset = 0;
  const flush = (end) => {
    if (start < 0) return;
    const raw = text.slice(start, end);
    const norm = raw.toLowerCase().replace(/['’]/g, "");
    if (norm) tokens.push({ norm, start, end });
    start = -1;
  };
  for (const ch of chars) {
    if (WORD_CHAR.test(ch)) {
      if (start < 0) start = offset;
    } else {
      flush(offset);
    }
    offset += ch.length;
  }
  flush(text.length);
  return tokens;
}

export const words = (phrase) => tokenize(phrase).map((t) => t.norm);
export const normalizedPhrase = (phrase) => words(phrase).join(" ");

/** Original text from token `index` to the end, trimmed. null if nothing left. */
export function remainder(text, tokens, index) {
  if (index >= tokens.length) return null;
  const rest = text.slice(tokens[index].start).trim();
  return rest || null;
}

/** Original text covering tokens[from] … tokens[to - 1]. */
export function span(text, tokens, from, to) {
  return text.slice(tokens[from].start, tokens[to - 1].end);
}

/** Matches multi-word phrases against a token list, longest phrase first. */
export class PhraseMatcher {
  /** @param {{value: string, phrases: string[]}[] | string[]} items */
  constructor(items) {
    const list = items.map((item) => (typeof item === "string" ? { value: item, phrases: [item] } : item));
    const entries = [];
    for (const item of list) {
      for (const phrase of item.phrases) {
        const w = words(phrase);
        if (w.length) entries.push({ words: w, value: item.value });
      }
    }
    this.entries = entries.sort((a, b) => b.words.length - a.words.length);
  }

  match(tokens, index) {
    for (const entry of this.entries) {
      const end = index + entry.words.length;
      if (end > tokens.length) continue;
      let ok = true;
      for (let i = 0; i < entry.words.length; i += 1) {
        if (tokens[index + i].norm !== entry.words[i]) { ok = false; break; }
      }
      if (ok) return { value: entry.value, length: entry.words.length };
    }
    return null;
  }

  matchesWhole(tokens) {
    const m = this.match(tokens, 0);
    return Boolean(m) && m.length === tokens.length;
  }
}

/** Indexed matcher helper: phrases grouped by position in `table`, value = String(index). */
export function indexedMatcher(table) {
  return new PhraseMatcher(table.map((row, index) => ({ value: String(index), phrases: row.phrases })));
}

File: Packages/VoxCore/Sources/VoxCore/Parsing/Tokenizer.swift
import Foundation

/// A word from an utterance: its normalized form for matching, and where it
/// sits in the original text so prompts can be passed on exactly as spoken.
public struct Token: Equatable, Sendable {
    public let norm: String
    public let range: Range<String.Index>
}

public enum Tokenizer {
    /// Splits on anything that is not a letter, digit or apostrophe, then
    /// lowercases and drops apostrophes: "What's up-to-date?" -> whats, up, to, date.
    public static func tokenize(_ text: String) -> [Token] {
        var tokens: [Token] = []
        var start: String.Index?
        var index = text.startIndex

        func flush(until end: String.Index) {
            guard let s = start else { return }
            let norm = String(text[s..<end].lowercased().filter { $0 != "'" && $0 != "\u{2019}" })
            if !norm.isEmpty {
                tokens.append(Token(norm: norm, range: s..<end))
            }
            start = nil
        }

        while index < text.endIndex {
            let ch = text[index]
            let isWordChar = ch.isLetter || ch.isNumber || ch == "'" || ch == "\u{2019}"
            if isWordChar {
                if start == nil { start = index }
            } else {
                flush(until: index)
            }
            index = text.index(after: index)
        }
        flush(until: text.endIndex)
        return tokens
    }

    /// Normalized words of a phrase, for matching against tokens.
    public static func words(_ phrase: String) -> [String] {
        tokenize(phrase).map(\.norm)
    }

    /// Single-string form of a phrase ("Free-Buff" -> "free buff").
    public static func normalizedPhrase(_ phrase: String) -> String {
        words(phrase).joined(separator: " ")
    }

    /// Original text from token `index` to the end, trimmed. nil if nothing left.
    public static func remainder(of text: String, tokens: [Token], from index: Int) -> String? {
        guard index < tokens.count else { return nil }
        let rest = text[tokens[index].range.lowerBound...]
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return rest.isEmpty ? nil : rest
    }
}

/// Matches multi-word phrases against a token list.
public struct PhraseMatcher: Sendable {
    /// (words, value) pairs, longest phrases first so "claude code" beats "claude".
    private let entries: [(words: [String], value: String)]

    public init(_ phrasesByValue: [(value: String, phrases: [String])]) {
        var entries: [(words: [String], value: String)] = []
        for item in phrasesByValue {
            for phrase in item.phrases {
                let w = Tokenizer.words(phrase)
                if !w.isEmpty { entries.append((w, item.value)) }
            }
        }
        self.entries = entries.sorted { $0.words.count > $1.words.count }
    }

    /// Convenience for plain phrase lists where the value is the phrase itself.
    public init(phrases: [String]) {
        self.init(phrases.map { (value: $0, phrases: [$0]) })
    }

    /// Longest phrase starting at `index`: its value and how many tokens it used.
    public func match(_ tokens: [Token], at index: Int) -> (value: String, length: Int)? {
        for entry in entries {
            let end = index + entry.words.count
            guard end <= tokens.count else { continue }
            if zip(tokens[index..<end], entry.words).allSatisfy({ pair in pair.0.norm == pair.1 }) {
                return (entry.value, entry.words.count)
            }
        }
        return nil
    }

    /// True if the whole token list is exactly one of the phrases.
    public func matchesWhole(_ tokens: [Token]) -> Bool {
        guard let m = match(tokens, at: 0) else { return false }
        return m.length == tokens.count
    }
}
